lib/matplotlib/units.py

```diff
@@ -43,12 +43,85 @@ def default_units(x, axis):
 """
 
 from numbers import Number
+import inspect
+import functools
 
 import numpy as np
 
 from matplotlib.cbook import iterable, safe_first_element
 
 
+def _accepts_units(convert_x=[], convert_y=[]):
+    """
+    A decorator for functions and methods that accept units. The parameters
+    indicated in *convert_x* and *convert_y* are used to update the axis
+    unit information, are converted, and then handed on to the decorated
+    function.
+
+    Parameters
+    ----------
+    convert_x, convert_y : list
+        A list of integers or strings, indicating the arguments to be converted
+    """
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            axes = args[0]
+            # Bind the incoming arguments to the function signature
+            bound_args = inspect.signature(func).bind(*args, **kwargs)
+            bound_args.apply_defaults()
+            # Get the original arguments - these will be modified later
+            arguments = bound_args.arguments
+            # Check for data kwarg
+            has_data = (('data' in arguments) and
+                        (arguments['data'] is not None))
+            if has_data:
+                data = arguments['data']
+
+            # Helper method to process unit info, and convert *original_data*
+            def _process_info(original_data, axis):
+                if original_data is None:
+                    return
+                if axis == 'x':
+                    axes._process_unit_info(xdata=original_data, kwargs=kwargs)
+                    converted_data = axes.convert_xunits(original_data)
+                elif axis == 'y':
+                    axes._process_unit_info(ydata=original_data, kwargs=kwargs)
+                    converted_data = axes.convert_yunits(original_data)
+                return converted_data
+
+            # Loop through each argument to be converted, update the axis
+            # unit info, convert argument, and replace in *arguments* with
+            # converted values
+            for arg in convert_x:
+                if has_data and arguments[arg] in data:
```

Comment:
This does not work for structured numpy arrays:

~~~
>>> a = np.array([(1, 2), (3, 4)], dtype=[('x', float), ('y', float)])
>>> 'x' in a
False
>>> print(a['x'])
[1, 3]
~~~

We've had the same issue/discussion a while ago in a similar context. Unfortunately, I cannot find it anymore. Maybe someone remembers?

Reply:
Do you know what works instead? Do we officially support structured numpy arrays?

Reply:
See the ` _has_item()` function in #10872 for a working check.

https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.plot.html says:

> All indexable objects are supported. This could e.g. be a dict, a pandas. DataFame or a structured numpy array.

Not sure if I have written it in there with the last changes. But at least nobody has vetoed.